### ISSUE TITLE
bug

### ISSUE THREAD

**Issue by mf85** [zemasoft/eset-challenge-evm2#3]
Hey,
When I run the attached program , virtual machine crashes.

[crash2.zip](https://github.com/zemasoft/eset-challenge-evm2/files/2668462/crash2.zip)

**Comment by zemasoft** [comment]
Great! Thank you for reporting this. I love bug reports :-)

I tried to run the program and the VM really crashed. I was even unable to list the program (using -L switch) so without going any deeper, I think the file is probably corrupted. I know that not all cases of corruption are currently handled.

Did you corrupt the file yourself? Or is the compiler wrong?

If I'm totally mistaken and the file is OK, can you please provide me with the original source? And is your VM able to run this?

Thanks!

**Comment by mf85** [comment]
Hey,
malicious intent well spotted :-)
I corrupted the file , this is an Eset task so I thought some hacking is ok :-P

Do you want the source or would you rather reverse the file yourself?
Just two instructions :-)

**Comment by zemasoft** [comment]
Now when I think about it, this is still an incorrect behavior (but a different issue) since there is an incomplete mov instruction at the end of the file which is just ignored by the VM. It should at least be reported. Maybe this is the thing I had in mind.

**Comment by zemasoft** [comment]
Note you can list the program with:

```bash
$ ./build/app/evm2 -u -L crash2.evm
```

where `-u` stands for `unresolved`. The VM will leave the addresses unresolved when this switch is used.

**Comment by mf85** [comment]
Tested. It works although one more thing.
The attached program is a variation of the same problem, but this time the corrupted address points inside the code segment. Your VM will not run it and exits with "buildtime error: Unresolvable address".
My VM will run it like this:
00000000: jump 45h
00000045: mov r0, r0
00000052: mov r0, r0
0000005f: mov r0, r0
0000006c: PADDING

Of course the specification says nothing about how to handle situations like this, but it is interesting how the lack of a specification can lead to slightly different implementations.

Anyway, thx for the opportunity to have some fun with a little bit of "hacking" :-)

[invalid_jmp.zip](https://github.com/zemasoft/eset-challenge-evm2/files/2671309/invalid_jmp.zip)

**Comment by zemasoft** [comment]
I think this lack of specification is there with purpose, to have diverse implementations and rich discussions :-)

So you are really moving your instruction pointer within the bitcode (and even further). My design decision was to "build" the code first to have the solution more performant. Your solution, on the other hand, is more versatile, since you should be, for example, able to jump in the middle of an instruction. You can try to tweak the compiler for that. I think the syntax back than in tasm was like: `jmp $+10` - jump to this instruction address + 10.

**Comment by mf85** [comment]
Yep, I have an IP register. I also thought about cache-ing the instructions so the VM does not have to decode them all the time but I was lazy to do it. Instead I have made a "just in time" recompiler that translates the bitcode into a x64 machine code, but just like your solution it cannot handle invalid jumps because it translates the program as a whole. Probably the right solution is to translate/"build" the code in blocks just like QEMU does it. 
https://www.slideshare.net/RampantJeff/qemu-binary-translation
from slide 12 if you are interested
